das/google_cloud_storage_service.go

```diff
@@ -0,0 +1,158 @@
+package das
+
+import (
+	googlestorage "cloud.google.com/go/storage"
+	"context"
+	"encoding/hex"
+	"fmt"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/google/go-cmp/cmp"
+	"github.com/offchainlabs/nitro/arbstate/daprovider"
+	"github.com/offchainlabs/nitro/das/dastree"
+	"github.com/offchainlabs/nitro/util/pretty"
+	flag "github.com/spf13/pflag"
+	"google.golang.org/api/option"
+	"io"
+	"sort"
+)
+
+type GoogleCloudStorageOperator interface {
+	Bucket(name string) *googlestorage.BucketHandle
+	Upload(ctx context.Context, bucket, objectPrefix string, value []byte) error
+	Download(ctx context.Context, bucket, objectPrefix string, key common.Hash) ([]byte, error)
+	Close(ctx context.Context) error
+}
+
+type GoogleCloudStorageClient struct {
+	client *googlestorage.Client
+}
+
+func (g *GoogleCloudStorageClient) Bucket(name string) *googlestorage.BucketHandle {
+	return g.client.Bucket(name)
+}
+
+func (g *GoogleCloudStorageClient) Upload(ctx context.Context, bucket, objectPrefix string, value []byte) error {
+	obj := g.client.Bucket(bucket).Object(objectPrefix + EncodeStorageServiceKey(dastree.Hash(value)))
+	w := obj.NewWriter(ctx)
+
+	if _, err := fmt.Fprintln(w, hex.EncodeToString(value)); err != nil {
+		return err
+	}
+	return w.Close()
+
+}
+
+func (g *GoogleCloudStorageClient) Download(ctx context.Context, bucket, objectPrefix string, key common.Hash) ([]byte, error) {
+	obj := g.client.Bucket(bucket).Object(objectPrefix + EncodeStorageServiceKey(key))
+	reader, err := obj.NewReader(ctx)
+	if err != nil {
+		return nil, err
+	}
+	return io.ReadAll(reader)
+}
+
+func (g *GoogleCloudStorageClient) Close(ctx context.Context) error {
+	return g.client.Close()
+}
+
+type GoogleCloudStorageServiceConfig struct {
+	Enable              bool   `koanf:"enable"`
+	AccessToken         string `koanf:"access-token"`
+	Bucket              string `koanf:"bucket"`
+	ObjectPrefix        string `koanf:"object-prefix"`
+	DiscardAfterTimeout bool   `koanf:"discard-after-timeout"`
```

Comment:
I prefer the name `enable-expiry` now, I'm using it in local_file_storage_service.go and I think it's clearer.

It looks like expiry isn't really supported yet by this implementation so either just leave this option out entirely, or add expiry support.

Reply:
Yes, both enable-expiry and max-retention sounds more clear to me too!